Modules/_heapqmodule.c

```diff
@@ -538,6 +538,460 @@ _heapq__heapify_max(PyObject *module, PyObject *heap)
+
+    return (PyObject *) mo;
+
+error:
+    Py_XDECREF(keyfunc);
+    Py_XDECREF(iterators);
+    Py_XDECREF(tree);
+    return NULL;
+}
+
+static void
+merge_dealloc(mergeobject *mo)
+{
+    PyObject_GC_UnTrack(mo);
+    Py_XDECREF(mo->tree);
+    Py_XDECREF(mo->iterators);
+    Py_XDECREF(mo->keyfunc);
+    Py_TYPE(mo)->tp_free(mo);
+}
+
+static PyObject *
+merge_sizeof(mergeobject *mo, void *unused)
+{
+    Py_ssize_t res = _PyObject_SIZE(Py_TYPE(mo));
+
+    if (mo->iterators)
+        res += PyList_GET_SIZE(mo->iterators) * sizeof(PyObject *);
+    if (mo->tree)
+        res += PyList_GET_SIZE(mo->tree) * sizeof(PyObject *);
+    return PyLong_FromSsize_t(res);
+}
+
+static int
+merge_traverse(mergeobject *mo, visitproc visit, void *arg)
+{
+    Py_VISIT(mo->tree);
+    Py_VISIT(mo->iterators);
+    Py_VISIT(mo->keyfunc);
+    return 0;
+}
+
+static PyObject *
+merge_next(mergeobject *mo) {
+    PyObject *result;
+    if (!mo->started) {
+        int i;
+        int n = PyList_GET_SIZE(mo->iterators);
+        /* Heapify, except leaves are supplied by iterators. */
+        if (mo->keyfunc == NULL) {
+            for (i = 2 * n - 2; i > 0; i--) {
+                if (_tree_sift(mo, i) < 0)
+                    return NULL;
+            }
+        }
+        else {
+            for (i = 4 * n - 4; i > 0; i -= 2) {
+                if (_tree_sift_key(mo, i) < 0)
+                    return NULL;
+            }
+        }
+        mo->started = 1;
+    }
+    if (mo->tree == NULL)
+        return NULL;
+    if (mo->keyfunc == NULL) {
+        if (_tree_sift(mo, 0) < 0)
+            return NULL;
+        result = PyList_GET_ITEM(mo->tree, 0);
+        PyList_SET_ITEM(mo->tree, 0, NULL);
+    }
+    else {
+        if (_tree_sift_key(mo, 0) < 0)
+            return NULL;
+        Py_XDECREF(PyList_GET_ITEM(mo->tree, 0));
+        PyList_SET_ITEM(mo->tree, 0, NULL);
+        result = PyList_GET_ITEM(mo->tree, 1);
+        PyList_SET_ITEM(mo->tree, 1, NULL);
+    }
+    return result;
+}
+
+
+PyDoc_STRVAR(sizeof_doc, "Returns size in memory, in bytes.");
+
+static PyMethodDef merge_methods[] = {
+    {"__sizeof__", (PyCFunction)merge_sizeof, METH_NOARGS, sizeof_doc},
+    {NULL, NULL}
+};
+
+PyDoc_STRVAR(merge_doc,
+"merge(*iterables, key=None, reverse=False) --> merge object\n\
+\n\
+Merge multiple sorted inputs into a single sorted output.\n\
+\n\
+Similar to sorted(itertools.chain(*iterables)) but returns an iterator,\n\
+does not pull the data into memory all at once, and assumes that each of\n\
+the input streams is already sorted (smallest to largest).\n\
+\n\
+>>> list(merge([1,3,5,7], [0,2,4,8], [5,10,15,20], [], [25]))\n\
+[0, 1, 2, 3, 4, 5, 5, 7, 8, 10, 15, 20, 25]\n\
+\n\
+If *key* is not None, applies a key function to each element to determine\n\
+its sort order.\n\
+\n\
+>>> list(merge(['dog', 'horse'], ['cat', 'fish', 'kangaroo'], key=len))\n\
+['dog', 'cat', 'fish', 'horse', 'kangaroo']");
+
+static PyTypeObject merge_type = {
```

Comment:
Please,  use C99 designated initializers for this class as is new (we use C99 now :) ).